RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,594 @@
+                                      dummy,
+                                      theClusterParam.sx2,
+                                      dummy);
+
+  theClusterParam.sigmax = theClusterParam.sigmax * micronsToCm;
+  theClusterParam.sx1 = theClusterParam.sx1 * micronsToCm;
+  theClusterParam.sx2 = theClusterParam.sx2 * micronsToCm;
+
+  theClusterParam.sigmay = theClusterParam.sigmay * micronsToCm;
+  theClusterParam.sy1 = theClusterParam.sy1 * micronsToCm;
+  theClusterParam.sy2 = theClusterParam.sy2 * micronsToCm;
+}
+
+//-----------------------------------------------------------------------------
+//! Hit position in the local frame (in cm).  Unlike other CPE's, this
+//! one converts everything from the measurement frame (in channel numbers)
+//! into the local frame (in centimeters).
+//-----------------------------------------------------------------------------
+LocalPoint PixelCPEFast::localPosition(DetParam const& theDetParam, ClusterParam& theClusterParamBase) const {
+  ClusterParamGeneric& theClusterParam = static_cast<ClusterParamGeneric&>(theClusterParamBase);
+
+  assert(!theClusterParam.with_track_angle);
+
+  if (UseErrorsFromTemplates_) {
+    errorFromTemplates(theDetParam, theClusterParam, theClusterParam.theCluster->charge());
+  } else {
+    theClusterParam.qBin_ = 0;
+  }
+
+  int Q_f_X;  //!< Q of the first  pixel  in X
+  int Q_l_X;  //!< Q of the last   pixel  in X
+  int Q_f_Y;  //!< Q of the first  pixel  in Y
+  int Q_l_Y;  //!< Q of the last   pixel  in Y
+  collect_edge_charges(theClusterParam, Q_f_X, Q_l_X, Q_f_Y, Q_l_Y, UseErrorsFromTemplates_ && TruncatePixelCharge_);
+
+  // do GPU like ...
+  pixelCPEforGPU::ClusParams cp;
+
+  cp.minRow[0] = theClusterParam.theCluster->minPixelRow();
+  cp.maxRow[0] = theClusterParam.theCluster->maxPixelRow();
+  cp.minCol[0] = theClusterParam.theCluster->minPixelCol();
+  cp.maxCol[0] = theClusterParam.theCluster->maxPixelCol();
+
+  cp.Q_f_X[0] = Q_f_X;
+  cp.Q_l_X[0] = Q_l_X;
+  cp.Q_f_Y[0] = Q_f_Y;
+  cp.Q_l_Y[0] = Q_l_Y;
+
+  auto ind = theDetParam.theDet->index();
+  pixelCPEforGPU::position(m_commonParamsGPU, m_detParamsGPU[ind], cp, 0);
+  auto xPos = cp.xpos[0];
+  auto yPos = cp.ypos[0];
+
+  //  std::cout<<" in PixelCPEFast:localPosition - pos = "<<xPos<<" "<<yPos
+  //           << " size "<< cp.maxRow[0]-cp.minRow[0] << ' ' << cp.maxCol[0]-cp.minCol[0] << std::endl; //dk
+
+  //--- Now put the two together
+  LocalPoint pos_in_local(xPos, yPos);
+  return pos_in_local;
+}
+
+//-----------------------------------------------------------------------------
+//!  Collect the edge charges in x and y, in a single pass over the pixel vector.
+//!  Calculate charge in the first and last pixel projected in x and y
+//!  and the inner cluster charge, projected in x and y.
+//-----------------------------------------------------------------------------
+void PixelCPEFast::collect_edge_charges(ClusterParam& theClusterParamBase,  //!< input, the cluster
```

Comment:
looks like a copy-paste from PixelCPEGeneric -> reuse?

Reply:
Added to the post-merge open issues at https://github.com/cms-sw/cmssw/issues/32483 .